linkerd/app/core/src/proxy/server.rs

```diff
@@ -223,17 +220,14 @@ where
                             .serve_connection(io, HyperServerSvc::new(http_svc));
 
                         tokio::select! {
-                            res = &mut conn => {
-                                res?
-                            }
+                            res = &mut conn => { res }
                             () = drain.into_future() => {
-                                Pin::new(&mut conn).graceful_shutdown()
+                                Pin::new(&mut conn).graceful_shutdown();
+                                conn.await
```

Comment:
I don't think this is gonna cut it, either. We still need the future _triggering_ the graceful shutdown (which is in the runtime's `block_on` future) to wait for all the graceful shutdowns to complete, which is what `drain::Sender::drained()` did. Not waiting in `block_on` will mean the runtime can exit before the drain has completed.